service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/twin/TwinClient.java

```diff
@@ -256,13 +276,20 @@ public void patch(Twin twin) throws IotHubException, IOException
             url = IotHubConnectionString.getUrlModuleTwin(this.hostName, twin.getDeviceId(), twin.getModuleId());
         }
 
-        TwinState twinState = new TwinState(twin.getTagsMap(), twin.getDesiredMap(), null);
+        TwinState twinState = new TwinState(twin.getTags(), twin.getDesiredProperties(), null);
         String twinJson = twinState.toJsonElement().toString();
 
         HttpRequest httpRequest = createRequest(url, HttpMethod.PATCH, twinJson.getBytes(StandardCharsets.UTF_8));
 
+        if (ifMatch != null)
+        {
+            httpRequest.setHeaderField(IF_MATCH, ifMatch);
+        }
+
         // no need to return http response since method returns void
-        httpRequest.send();
+        HttpResponse httpResponse = httpRequest.send();
+        String twinString = new String(httpResponse.getBody(), StandardCharsets.UTF_8);
+        return Twin.fromJson(twinString);
```

Comment:
Instead of returning void, we'll return the new twin as returned from the service